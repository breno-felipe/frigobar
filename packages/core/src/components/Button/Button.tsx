import React from 'react';
import styled, { DefaultTheme } from 'styled-components';

interface ButtonProps {
  children: React.ReactNode;
  /** Change background-color */
  skin?: 'primary' | 'info' | 'success' | 'danger' | 'warning' | 'neutral';
  /** change the width, height and font-size values */
  size?: 'small' | 'medium' | 'large' | undefined;
  /** Change disabled state of the button */
  disabled?: boolean;
  /** 100% width */
  full?: boolean;
  /** add border-radius circle */
  rounded?: boolean;
  /** large button */
  large?: boolean;
}

const sizes = ({ spacings }: DefaultTheme) => ({
  small: {
    font: 0.875,
    icon: 1,
    padding: {
      top: spacings.xxsmall,
      right: spacings.small,
      bottom: spacings.xxsmall,
      left: spacings.small,
    },
  },
  medium: {
    font: 1,
    icon: 1.275,
    padding: {
      top: spacings.xsmall,
      right: spacings.medium,
      bottom: spacings.xsmall,
      left: spacings.medium,
    },
  },
  large: {
    font: 1.275,
    icon: 1.5,
    padding: {
      top: spacings.small,
      right: spacings.large,
      bottom: spacings.small,
      left: spacings.large,
    },
  },
});

const Button = styled.button<ButtonProps>`
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  transition: all 200ms ease-in-out;
  text-decoration: none;

  ${({
    theme,
    skin = 'neutral',
    size = 'small',
    disabled,
    full,
    rounded,
    large,
  }) => {
    const { radius, colors } = theme;

    return `
      background-color: ${colors[skin][500]};
      color: ${colors.neutral[50]};
      font-size: ${large ? '1.125' : sizes(theme)[size].font}rem;
      padding: ${
        large
          ? `${sizes(theme)[size].padding.top * 2}px
             ${sizes(theme)[size].padding.right * 2}px
             ${sizes(theme)[size].padding.bottom * 2}px
             ${sizes(theme)[size].padding.left * 2}px
          `
          : `${sizes(theme)[size].padding.top}px
             ${sizes(theme)[size].padding.right}px
             ${sizes(theme)[size].padding.bottom}px
             ${sizes(theme)[size].padding.left}px
          `
      };
  
      ${
        disabled
          ? `
        background-color: ${colors.neutral[100]};
        color: ${colors.neutral[400]};
        cursor: not-allowed;
      `
          : `
          &:hover {
            box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.4);
          }
          `
      }
  
      ${full ? 'width: 100%;' : ''}
      ${rounded ? `border-radius: ${radius[3]}px` : ''}
    `;
  }};
`;

Button.defaultProps = {
  skin: 'neutral',
  disabled: false,
  size: 'small',
  full: undefined,
  rounded: undefined,
  large: undefined,
};

export default Button;
