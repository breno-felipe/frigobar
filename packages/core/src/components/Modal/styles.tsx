import styled, { css } from 'styled-components';

import hexToRgb from '../../utils/hexToRgb';

const Backdrop = styled.div(
  ({ theme }) => css`
    position: fixed;
    inset: 0;

    z-index: 999;

    display: flex;
    align-items: center;
    justify-content: center;

    background-color: ${hexToRgb(theme.colors.black, 0.6)};
  `,
);

const Dialog = styled.section(
  ({ theme }) => css`
    position: relative;

    min-width: 200px;
    min-height: 200px;

    padding: ${theme.spacings.medium}px;

    border: ${theme.borders.tiny}px solid ${theme.colors.neutral[300]};
    border-radius: ${theme.radius[2]}px;

    background-color: ${theme.colors.neutral[50]};
  `,
);

const CloseButton = styled.button(
  ({ theme }) => css`
    position: absolute;
    top: ${theme.spacings.xsmall}px;
    right: ${theme.spacings.xsmall}px;

    display: flex;
    align-items: center;
    justify-content: center;

    width: 28px;
    height: 28px;

    margin: 0;
    padding: 0;

    cursor: pointer;

    border: none;
    border-radius: ${theme.radius[3]}px;

    background-color: ${theme.colors.neutral[50]};

    &:hover {
      background-color: ${theme.colors.neutral[200]};
    }
  `,
);

export { Backdrop, Dialog, CloseButton };
