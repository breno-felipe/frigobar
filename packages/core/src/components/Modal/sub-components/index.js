import Content from './Content';
import Footer from './Footer';
import Title from './Title';
import Divider from './Divider';

export { Content, Footer, Title, Divider };
