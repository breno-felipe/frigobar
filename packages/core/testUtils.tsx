import React from 'react';
import PropTypes from 'prop-types';
import {
  render,
  fireEvent,
  waitFor,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from './src/theme';

const Providers = ({ children }: { children?: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

Providers.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, fireEvent, waitFor };
