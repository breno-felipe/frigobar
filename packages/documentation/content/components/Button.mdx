---
title: Button
menu: components
---

Button component

```jsx react-live=true
<Button>Button</Button>
```
