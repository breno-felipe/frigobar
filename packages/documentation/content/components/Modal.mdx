---
title: Modal
menu: components
---

Modal component

```jsx react-live=true noInline
render(() => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <>
      <Button onClick={() => setIsOpen(true)}>Open modal</Button>
      <Modal open={isOpen} onClose={() => setIsOpen(false)}>
        <a href="#">link</a>
        <button>button</button>
        <input />
      </Modal>
    </>
  );
});
```
