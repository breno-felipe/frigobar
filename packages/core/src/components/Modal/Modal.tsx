import React, { useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';
import { X } from 'react-feather';
import { useFade } from '@frigobar/animation';

import portalContainer from '../Menu/portalContainer';
import useFocusTrap from '../../utils/useFocusTrap';

import { Backdrop, Dialog, CloseButton } from './styles';

const PORTAL_CONTAINER_NAME = 'frigobar-modal';

function Modal({
  children,
  onClose,
  onClickOutside,
  open,
}: {
  children: React.ReactNode;
  onClose: () => void;
  onClickOutside: () => void;
  open: boolean;
}): JSX.Element | null {
  const [{ animation, state }, toggleFade] = useFade({ startOnRender: open });
  const dialogRef = useRef<HTMLElement>(null);
  portalContainer(PORTAL_CONTAINER_NAME);

  const handleKeyDown = useFocusTrap(dialogRef, [state]);

  function handleClickOutside(event: MouseEvent) {
    if (!dialogRef.current?.contains(event.target as Node)) {
      onClose && onClose();
      onClickOutside && onClickOutside();
    }
  }

  useEffect(() => {
    if (open) {
      window.addEventListener('click', handleClickOutside);
    } else {
      window.removeEventListener('click', handleClickOutside);
    }

    toggleFade(open);

    return () => window.removeEventListener('click', handleClickOutside);
  }, [open]);

  return state
    ? createPortal(
        <Backdrop
          animation={animation}
          onKeyDown={handleKeyDown}
          onAnimationEnd={() => console.log('end')}
        >
          <Dialog ref={dialogRef}>
            <CloseButton onClick={onClose}>
              <X width={24} height={24} />
            </CloseButton>
            {children}
          </Dialog>
        </Backdrop>,
        document.querySelector(`#${PORTAL_CONTAINER_NAME}`) || document.body,
      )
    : null;
}

export default Modal;
