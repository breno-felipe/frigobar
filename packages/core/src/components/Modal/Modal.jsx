import styled from 'styled-components';

import { Content, Footer, Title, Divider } from './sub-components';
import defaultTheme from '../../theme';

const Modal = styled.section(
  ({ maxWidth, theme: { colors } }) => `
  border-radius: 0px;
  display: inline-block;
  max-width: ${maxWidth};
  box-shadow: 0px 2px 5px 0px ${colors.neutral[200]};
  width: 100%;
`,
);

Modal.Content = Content;
Modal.Footer = Footer;
Modal.Title = Title;
Modal.Divider = Divider;

Modal.defaultProps = {
  theme: defaultTheme,
};

export default Modal;
