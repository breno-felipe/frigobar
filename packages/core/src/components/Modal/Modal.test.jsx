import React from 'react';
import { render } from '../../../testUtils';

import Modal from './Modal';
import Button from '../Button';

describe('<Modal />', () => {
  describe('Snapshots', () => {
    const Component = () => (
      <Modal maxWidth="450px">
        <Modal.Title>Lorem Ipsum</Modal.Title>
        <Modal.Content>
          Lorem Ipsum is simply dummy text of the printing and typesetting
          industry. Lorem Ipsum has been the industrys standard dummy text ever
          since the 1500.
        </Modal.Content>
        <Modal.Footer>
          <Button skin="primary" style={{ marginRight: 10 }}>
            Ok
          </Button>
          <Button skin="danger">Fechar</Button>
        </Modal.Footer>
      </Modal>
    );

    it('should match snapshot', () => {
      const { container } = render(<Component />);
      expect(container).toMatchSnapshot();
    });
  });
});
