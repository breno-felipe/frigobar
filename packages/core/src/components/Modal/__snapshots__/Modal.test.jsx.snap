// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< /> Snapshots should match snapshot 1`] = `
.c2 {
  padding: 8px 16px 8px 16px;
  font-family: Arial,Helvetica,sans-serif;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px 16px 8px 16px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c1 {
  font-size: 1.5rem;
  padding: 8px 16px 8px 16px;
  margin: 0;
}

.c0 {
  border-radius: 0px;
  display: inline-block;
  max-width: 450px;
  box-shadow: 0px 2px 5px 0px #ccc;
  width: 100%;
}

.c4 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #598aed;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c4:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

.c4 .c6 {
  font-size: 1rem;
}

.c5 {
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
  background-color: #f44336;
  color: #fdfdfd;
  font-size: 0.875rem;
  padding: 4px 12px 4px 12px;
}

.c5:hover {
  box-shadow: 0px 1px 3px 0px rgba(0,0,0,0.4);
}

.c5 .c6 {
  font-size: 1rem;
}

<div>
  <section
    class="c0"
  >
    <h2
      class="c1"
    >
      Lorem Ipsum
    </h2>
    <div
      class="c2"
    >
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500.
    </div>
    <footer
      class="c3"
    >
      <button
        class="c4"
        style="margin-right: 10px;"
      >
        Ok
      </button>
      <button
        class="c5"
      >
        Fechar
      </button>
    </footer>
  </section>
</div>
`;
